femtools/Detector_Tools.F90

```diff
@@ -643,65 +742,324 @@ subroutine set_particle_attribute_from_python(attributes, positions, nparts, fun
     !! the interpolated value of "FieldName" at the particle position at the current and previous time levels, and fields["OldAttributeName"] stores the attribute
-    !! value at the previous time level. The result must be a float. 
+    !! value at the previous time level. The result must be a float.
+    character(len=*), intent(in) :: func
+    !! Current model time
+    real, intent(in) :: time
+    !! Current model timestep
+    real, intent(in) :: dt
+    !! Whether this is an array-valued attribute
+    logical, intent(in) :: is_array
+
+    ! locals
+    integer :: i, j, field_idx
+    integer :: dim, stat, phase
+    real, dimension(:), pointer :: lvx, lvy, lvz
+    real, dimension(0), target :: zero
+    type(scalar_field), pointer :: sfield
+    type(vector_field), pointer :: vfield
+    type(tensor_field), pointer :: tfield
+    real, allocatable, dimension(:,:) :: field_vals, old_field_vals
+    type(detector_type), pointer :: particle
+
+    dim = size(positions(:,1))
+    select case(dim)
+    case(1)
+      lvx=>positions(1,:)
+      lvy=>zero
+      lvz=>zero
+    case(2)
+      lvx=>positions(1,:)
+      lvy=>positions(2,:)
+      lvz=>zero
+    case(3)
+      lvx=>positions(1,:)
+      lvy=>positions(2,:)
+      lvz=>positions(3,:)
+    end select
+
+    ! allocate space to hold fields and old fields
+    allocate(field_vals(field_counts(1) + dim*field_counts(2) + dim**2*field_counts(3), nparts))
+    allocate(old_field_vals(old_field_counts(1) + dim*old_field_counts(2) + dim**2*old_field_counts(3), nparts))
+
+    call evaluate_particle_fields(nparts, state, ele, lcoords, &
+         particle_list%field_names, particle_list%field_phases, field_counts, &
+         field_vals, dim)
+
+    ! copy old fields off particles
+    particle => particle_list%first
+    do i = 1, nparts
+      old_field_vals(:,i) = particle%old_fields(:)
+      particle => particle%next
+    end do
+
+    call set_scalar_particles_from_python_fields(func, len(func), dim, nparts, natt, &
+         lvx, lvy, lvz, time, dt, field_counts, field_names, field_vals, old_field_counts, old_field_names, &
+         old_field_vals, old_attr_counts, old_attr_names, old_attr_dims, old_attributes, is_array, attributes, stat)
+    if (stat/=0) then
+       ewrite(-1, *) "Python error, Python string was:"
+       ewrite(-1 , *) trim(func)
+       FLExit("Dying")
+    end if
+
+    deallocate(field_vals)
+    deallocate(old_field_vals)
+  end subroutine set_particle_scalar_attribute_from_python_fields
+
+  !> Given the particle position and time, evaluate the specified python function for a group of particles
+  !! specified in the string func at that location.
+  subroutine set_particle_vector_attribute_from_python(attributes, positions, nparts, natt, func, time, dt, is_array)
+    !! Attribute values to set
+    real, dimension(:,:), intent(out) :: attributes
```

Comment:
I was expecting a rank-3 array here: dim X natt X nparts

Reply:
All the attributes on a particle list are stored in a 2D array `(total att x nparts)`, so while I think it could make more sense to pass a 3D array into this function, it's a little more awkward.